### ISSUE TITLE
Unused Component/Resource Warning

### ISSUE THREAD

**Issue by Braymatter** [bevyengine/bevy#9776]
## What problem does this solve or what need does it fill?

It would be very nice if we had some tool that would emit a warning whenever a component or a resource is not a part of any query in the application. 

Sometimes this scenario can result in hard to track down bugs.

We have a DespawnAfterDuration component / system we use. After moving the parent plugin around to a subcrate, but not remounting the plugin to the app, the component never gets processed as the system/plugin aren't added to the app builder. 

Often times when moving code around / refactoring we'll forget to add a plugin / system etc. and we run into these sorts of things. Also very common for me when adding entirely new plugins / systems.

## What solution would you like?

Ideally a clippy lint  / dead_code would be best I think.

## What alternative(s) have you considered?

I've attempted to just be better, but it hasn't worked out yet.

**Comment by nicopap** [comment]
A lead would be to accumulate the `FilteredAccess` of all the schedules registered in the `App`, `OR`-ing the `write`, `read-write`, `with`, `without` fields, and check if there is any `0`s in the resulting bitset.

**Comment by hymm** [comment]
> A lead would be to accumulate the FilteredAccess of all the schedules registered in the App, OR-ing the write, read-write, with, without fields, and check if there is any 0s in the resulting bitset.

I don't think this will work. Components are registered when they're spawned or used in systems.

**Comment by hymm** [comment]
This is the whole `life before main` problem that needs support from rust to fix nicely.

The only way checking the schedule could work is if the user manually calls `init_component` on the world.

**Comment by tormeh** [comment]
Some related discussion: 
- https://github.com/rust-lang/rust-clippy/issues/6872
- https://github.com/bevyengine/bevy/issues/1602
